---
output: github_document
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# wbData


Easily download and use Wormbase data. Can be used to download or read the "geneIDs" table or the GTF files from Wormbase, and provides functions to convert between Wormbase IDs and gene symbols.

The three advantages are:

* finding files for any given Wormbase release (even though the actual path changes)
* keeping all these files in a single place
* facilitating common operations.

## Installation

You can install the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("AlexWeinreb/wbData")
```
## Example of gene ID conversion

To load the geneIDs table for Wormbase release WS273 and convert between gene ID and gene symbol:

```{r}
library(wbData)

gids <- wb_load_gene_ids("WS273")
s2i("unc-10", gids)
i2s(c("WBGene00006752", "WBGene00004412"), gids)
```


## Example reading gene coordinates

```{r}
gene_coords <- wbData::wb_load_gene_coords(273)
head(gene_coords)
```

This can be used in conjunction with other packages, for example to look up a gene model in IGV or JBrowse:

```{r, eval=FALSE}

gene_coords |>
  dplyr::filter(gene_id == s2i("unc-10", gids)) |>
  dplyr::pull(position) |>
  clipr::write_clip()

clipr::read_clip()
```

## Reading the CGC strain list

You can load the list of strains available at the CGC. The downloaded list is saved in cache, and by default is only downloaded again if the cached file is more than 2 days old. Use the `refresh` argument to force refreshing.

```{r}
strain_list <- wb_load_cgc_list()
nrow(strain_list)
```

For example, looking at a specific strain:

```{r}
strain_list$Genotype[strain_list$Strain == "NC902"]
```

Or looking for strains that have particular characteristics. For example, using a [regular expression](https://cran.r-project.org/web/packages/stringr/vignettes/regular-expressions.html) to look for strains that have *hbl-1* potentially associated with a red fluorescent protein:

```{r}

fluo_keywords <- c("cherry", "tomato", "scarlet", "rfp", " red ")
my_pattern <- paste("hbl-1", ".*::.*", fluo_keywords, sep="", collapse = "|")
with_fluo <- which(grepl(my_pattern, tolower(strain_list$Genotype)) | grepl(my_pattern, tolower(strain_list$Description)))
strain_list[with_fluo,]
```



## Data management

The downloaded file is stored on the local computer to avoid re-downloading every time. That way, `wbData` can also be used to obtain Wormbase files to use with other software.

The cache directory can be specified in three ways:

* Explicitly by specifying a path as argument to the package functions.
* Through `option("wb_dir_cache")`. In that case leave the argument as `NULL`.
* If the `dir_cache` argument is `NULL` and the option is not specified, a user-specific cache directory is chosen based on the operating system by [rappdirs](https://github.com/r-lib/rappdirs/).


To only list the files in cache without deleting them:

```{r}
wb_clean_cache(273, delete = FALSE)
```

The cache can be emptied with:

```{r}
wb_clean_cache(273)
```



